### PR TITLE
WinForms breaking changes for .NET 9 Preview 1

### LINKED ISSUE #37547: [Breaking change]: Nullability Change on IWindowsFormsEditorService.DropDownControl
### Description

The parameter on [IWindowsFormsEditorService.DropDownControl](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.design.iwindowsformseditorservice.dropdowncontrol?view=windowsdesktop-7.0) was previously marked as nullable, but we are changing it to be non-nullable via https://github.com/dotnet/winforms/pull/10088 as logically it should not take null and a NullReferenceException was discovered in some implementations of it while making nullability annotation changes.

### Version

.NET 9 Preview 1

### Previous behavior

`DropDownControl` method parameter is marked as nullable

### New behavior

`DropDownControl` method parameter no longer marked as nullable

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [X] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [ ] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

The parameter on `IWindowsFormsEditorService.DropDownControl` was previously marked as nullable, but docs do not describe to implementers how they should handle null input and logically this should not take null so API should not encourage passing null to this method with allowing nullable.

### Recommended action

Ensure that a nullable `Control` is not being passed as a parameter to `IWindowsFormsEditorService.DropDownControl` and any implementations of `IWindowsFormsEditorService` should update to remove nullability of parameter on `DropDownControl` method.

### Feature area

Windows Forms

### Affected APIs

`IWindowsFormsEditorService` and implementers

---
[Associated WorkItem - 196555](https://dev.azure.com/msft-skilling/Content/_workitems/edit/196555)

### LINKED ISSUE #38283: [Breaking change]: DataGridViewHeaderCell.MouseXUnsharesRow No Longer Throws NRE
### Description

Previously we would be throwing a null reference exception in `DataGridViewHeaderCell.MouseDownUnsharesRow`, `DataGridViewHeaderCell.MouseEnterUnsharesRow`, `DataGridViewHeaderCell.MouseLeaveUnsharesRow`, and `DataGridViewHeaderCell.MouseUpUnsharesRow` if `DataGridViewHeaderCell.DataGridView` was null. This is unexpected behavior as it is not documented. We have updated code to simply return `false` if `DataGridView` is null via https://github.com/dotnet/winforms/pull/10158.

### Version

.NET 9 Preview 1

### Previous behavior

Previously, if `DataGridViewHeaderCell.DataGridView` was null, user would end up getting a null reference exception if one of the above methods was called.

### New behavior

These methods will no longer throw null reference exception if `DataGridViewHeaderCell.DataGridView` is null. Instead it will just return false if it is null.

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

Upon updating the file to enable nullable annotations, the possibility for a null reference exception was caught. This behavior would be unexpected to the user as this is not documented

### Recommended action

If users were relying on the code to throw null reference exception for any reason, users should change code accordingly as this will no longer throw and just return `false` now.

### Feature area

Windows Forms

### Affected APIs

[DataGridViewHeaderCell.MouseDownUnsharesRow](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.datagridviewheadercell.mousedownunsharesrow?view=windowsdesktop-7.0)
[DataGridViewHeaderCell.MouseEnterUnsharesRow](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.datagridviewheadercell.mouseenterunsharesrow?view=windowsdesktop-7.0)
[DataGridViewHeaderCell.MouseLeaveUnsharesRow](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.datagridviewheadercell.mouseleaveunsharesrow?view=windowsdesktop-7.0)
[DataGridViewHeaderCell.MouseUpUnsharesRow](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.datagridviewheadercell.mouseupunsharesrow?view=windowsdesktop-7.0)

---
[Associated WorkItem - 196556](https://dev.azure.com/msft-skilling/Content/_workitems/edit/196556)

### LINKED ISSUE #38363: [Breaking change]: BindingSource.SortDesciptions No Longer Returns Null
### Description

BindingSource.SortDescriptions has been updated to return an empty `ListSortDescriptionCollection` instead of `null` if the data source is not an `IBindingListView` via https://github.com/dotnet/winforms/pull/10253

### Version

.NET 9 Preview 1

### Previous behavior

`BindingSource.SortDescriptions` return type:

> If the data source is an [IBindingListView](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.ibindinglistview?view=windowsdesktop-7.0), a [ListSortDescriptionCollection](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.listsortdescriptioncollection?view=windowsdesktop-7.0) that contains the sort descriptions applied to the list; otherwise, null.

(https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.bindingsource.sortdescriptions?view=windowsdesktop-7.0)

### New behavior

`BindingSource.SortDescriptions` will now return an empty `ListSortDescriptionCollection` where is was returning `null` before

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

`BindingSource.SortDescriptions` has historically returned `null` if the data source was not an `IBindingListView`. However `BindingSource.SortDescriptions` implements [IBindingListView.SortDescriptions](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.ibindinglistview.sortdescriptions?view=net-8.0) , whose return is marked as nonnullable. This was discovered while enabling nullability in the class. In order to align with the interface it is implementing, the API has been changed to now return an empty `ListSortDescriptionCollection` instead of `null`

### Recommended action

If users have been expecting `null` from `BindingSource.SortDescriptions` for any reason, this should be changed to expect an empty `ListSortDescriptionCollection`

### Feature area

Windows Forms

### Affected APIs

[BindingSource.SortDescriptions](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.bindingsource.sortdescriptions?view=windowsdesktop-7.0)

---
[Associated WorkItem - 196553](https://dev.azure.com/msft-skilling/Content/_workitems/edit/196553)

### LINKED ISSUE #38527: [Breaking change]: DataGridViewRowAccessibleObject's Name field will start at Row 1 by default, instead of Row 0
### Description

In this update, the `DataGridViewRowAccessibleObject` undergoes a modification affecting the `Name` field. The row index  in the `Name` field now starts at 1 instead of 0 by default.

As a result of this change, Narrator will announce the selected row of a DataGridView with a starting index at 1.

To maintain the previous behavior and start the row index at 0, a new switch, `System.Windows.Forms.DataGridViewUIAStartRowCountAtZero`, has been introduced. Setting this switch to true ensures that the `DataGridViewRowAccessibleObject` displays the row index with a starting point at 0, restoring the original functionality.

### Version

.NET 9 Preview 1

### Previous behavior

Prior to this update, in the `DataGridViewRowAccessibleObject`, the `Name` field contained the row index with a starting point at 0.

Previously, Narrator would announce the selected row of a DataGridView with a starting index at 0.

### New behavior

With this change, the `DataGridViewRowAccessibleObject` now presents the row index in the `Name` field starting at 1.

Starting from this change, Narrator will announce the selected row of a DataGridView with a starting index at 1.

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

The modification addresses an accessibility concern highlighted in GitHub issue [#7154](https://github.com/dotnet/winforms/issues/7154). The issue pertains to the row counting in the DataGridViewRowAccessibleObject starting at 0, creating a discrepancy with user expectations and Narrator announcements.

The change ensures a more intuitive and inclusive experience for users relying on screen readers and accessibility tools while providing developers with the flexibility to maintain the original behavior if necessary.

### Recommended action

If your application relies on the previous behavior, and you prefer the row index to start at 0, utilize the newly introduced switch, `System.Windows.Forms.DataGridViewUIAStartRowCountAtZero`. Create a [runtimeconfig.template.json](https://github.com/dotnet/docs/files/13514519/runtimeconfig.template.json) file at the root folder of your project, setting this switch to true to maintain the original functionality. Update your codebase accordingly to accommodate this change and ensure that the `DataGridViewRowAccessibleObject` displays the row index with a starting point at 0.

Sample of a [runtimeconfig.template.json](https://github.com/dotnet/docs/files/13514519/runtimeconfig.template.json) file that utilizes the previous behavior:

```json
{
    "configProperties": {
      "System.Windows.Forms.DataGridViewUIAStartRowCountAtZero": true
    }
}
```

### Feature area

C#, Windows Forms

### Affected APIs
[DataGridViewRow.DataGridViewRowAccessibleObject.Name](https://learn.microsoft.com/dotnet/api/system.windows.forms.datagridviewrow.datagridviewrowaccessibleobject.name?view=windowsdesktop-8.0)

---
[Associated WorkItem - 196550](https://dev.azure.com/msft-skilling/Content/_workitems/edit/196550)

### LINKED ISSUE #38799: [Breaking change]: ComponentDesigner.Initialize Now Throws ArgumentNullException
### Description

`ComponentDesigner.Initialize` was updated to throw an `ArgumentNullException` if `null` is passed as a parameter via PR https://github.com/dotnet/winforms/pull/9408. While enabling nullability in the file, it was found that many methods/properties both in `ComponentDesigner` and subclassed objects relied on the passed in component to be initialized to non-null and would result in NRE or some other exception later during use if initialized with null value.

### Version

.NET 9 Preview 1

### Previous behavior

Previously passing `null` to `ComponentDesigner.Initialize` did not throw, but would result in NRE or some other exception during use in either `ComponentDesigner` or subclassed objects.

### New behavior

`ArgumentNullException` is now thrown if `null` is passed as an argument to `ComponentDesigner.Initialize`

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries may encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code may require source changes to compile successfully.
- [X] **Behavioral change**: Existing binaries may behave differently at run time.

### Reason for change

While enabling nullability in the file, it was found that many methods/properties both in `ComponentDesigner` and subclassed objects relied on the passed in component to be initialized to non-null and would result in NRE or other exceptions later during use if initialized with null value.

### Recommended action

Call `ComponentDesigner.Initialize` with nonnull parameter.

### Feature area

Windows Forms

### Affected APIs

[ComponentDesigner.Initialize](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.design.componentdesigner.initialize?view=windowsdesktop-8.0)

---
[Associated WorkItem - 199400](https://dev.azure.com/msft-skilling/Content/_workitems/edit/199400)

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #38799 
Fixes #38527 
Fixes #38363 
Fixes #38283 
Fixes #37547

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/compatibility/9.0.md](https://github.com/dotnet/docs/blob/3f3216a95dca99727dd3b8f1e2b3a3fd8703f187/docs/core/compatibility/9.0.md) | [Breaking changes in .NET 9](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/9.0?branch=pr-en-us-39188) |
| [docs/core/compatibility/windows-forms/9.0/componentdesigner-initialize.md](https://github.com/dotnet/docs/blob/3f3216a95dca99727dd3b8f1e2b3a3fd8703f187/docs/core/compatibility/windows-forms/9.0/componentdesigner-initialize.md) | [ComponentDesigner.Initialize throws ArgumentNullException](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/windows-forms/9.0/componentdesigner-initialize?branch=pr-en-us-39188) |
| [docs/core/compatibility/windows-forms/9.0/datagridviewheadercell-nre.md](https://github.com/dotnet/docs/blob/3f3216a95dca99727dd3b8f1e2b3a3fd8703f187/docs/core/compatibility/windows-forms/9.0/datagridviewheadercell-nre.md) | ["Breaking change: No exception if DataGridView is null"](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/windows-forms/9.0/datagridviewheadercell-nre?branch=pr-en-us-39188) |
| [docs/core/compatibility/windows-forms/9.0/datagridviewrowaccessibleobject-name-row.md](https://github.com/dotnet/docs/blob/3f3216a95dca99727dd3b8f1e2b3a3fd8703f187/docs/core/compatibility/windows-forms/9.0/datagridviewrowaccessibleobject-name-row.md) | [DataGridViewRowAccessibleObject.Name starting row index](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/windows-forms/9.0/datagridviewrowaccessibleobject-name-row?branch=pr-en-us-39188) |
| [docs/core/compatibility/windows-forms/9.0/nullability-changes.md](https://github.com/dotnet/docs/blob/3f3216a95dca99727dd3b8f1e2b3a3fd8703f187/docs/core/compatibility/windows-forms/9.0/nullability-changes.md) | [docs/core/compatibility/windows-forms/9.0/nullability-changes](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/windows-forms/9.0/nullability-changes?branch=pr-en-us-39188) |
| [docs/core/compatibility/windows-forms/9.0/sortdescriptions-return-value.md](https://github.com/dotnet/docs/blob/3f3216a95dca99727dd3b8f1e2b3a3fd8703f187/docs/core/compatibility/windows-forms/9.0/sortdescriptions-return-value.md) | [BindingSource.SortDescriptions doesn't return null](https://review.learn.microsoft.com/en-us/dotnet/core/compatibility/windows-forms/9.0/sortdescriptions-return-value?branch=pr-en-us-39188) |

### CODE REVIEW

File: docs/core/compatibility/windows-forms/9.0/componentdesigner-initialize.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+Starting in .NET 9, <xref:System.ComponentModel.Design.ComponentDesigner.Initialize%2A?displayProperty=nameWithType> throws an <xref:System.ArgumentNullException> if the argument is `null`.
+
+## Change category
+
+This change is a [*behavioral change*](../../categories.md#behavioral-change).
+
+## Reason for change
+
+When we enabled nullability in the code file, we found that many methods and properties, both in <xref:System.ComponentModel.Design.ComponentDesigner> and its subclasses, relied on the passed-in component to be initialized to non-`null`. These methods and properties resulted in a <xref:System.NullReferenceException> or another exception later on if they were initialized with a `null` value.
```
Comment (lonitra) [inline docs/core/compatibility/windows-forms/9.0/componentdesigner-initialize.md:28]: Should 'we' be avoided here?
```suggestion
During the process of enabling nullability in the code file, it was discovered that many methods and properties, both in <xref:System.ComponentModel.Design.ComponentDesigner> and its subclasses, relied on the passed-in component to be initialized to non-`null`. These methods and properties resulted in a <xref:System.NullReferenceException> or another exception later on if they were initialized with a `null` value.
```

File: docs/core/compatibility/windows-forms/9.0/datagridviewheadercell-nre.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+---
+title: "Breaking change: No exception if DataGridView is null"
+description: Learn about the breaking change in .NET 9 for Windows Forms where DataGridViewHeaderCell methods no longer throw an exception if the DataGridView property is null.
+ms.date: 01/16/2024
+---
+# No exception if DataGridView is null
+
+Previously, a <xref:System.NullReferenceException> was thrown in <xref:System.Windows.Forms.DataGridViewHeaderCell.MouseDownUnsharesRow(System.Windows.Forms.DataGridViewCellMouseEventArgs)?displayProperty=nameWithType>, <xref:System.Windows.Forms.DataGridViewHeaderCell.MouseEnterUnsharesRow(System.Int32)?displayProperty=nameWithType>, <xref:System.Windows.Forms.DataGridViewHeaderCell.MouseLeaveUnsharesRow(System.Int32)?displayProperty=nameWithType>, and <xref:System.Windows.Forms.DataGridViewHeaderCell.MouseUpUnsharesRow(System.Windows.Forms.DataGridViewCellMouseEventArgs)?displayProperty=nameWithType> if the <xref:System.Windows.Forms.DataGridViewElement.DataGridView> property was null. That behavior was unexpected and incorrect. We've updated these methods to simply return `false` if `DataGridView` is `null`.
```
Comment (lonitra) [inline docs/core/compatibility/windows-forms/9.0/datagridviewheadercell-nre.md:8]:
```suggestion
Previously, a <xref:System.NullReferenceException> was thrown in <xref:System.Windows.Forms.DataGridViewHeaderCell.MouseDownUnsharesRow(System.Windows.Forms.DataGridViewCellMouseEventArgs)?displayProperty=nameWithType>, <xref:System.Windows.Forms.DataGridViewHeaderCell.MouseEnterUnsharesRow(System.Int32)?displayProperty=nameWithType>, <xref:System.Windows.Forms.DataGridViewHeaderCell.MouseLeaveUnsharesRow(System.Int32)?displayProperty=nameWithType>, and <xref:System.Windows.Forms.DataGridViewHeaderCell.MouseUpUnsharesRow(System.Windows.Forms.DataGridViewCellMouseEventArgs)?displayProperty=nameWithType> if the <xref:System.Windows.Forms.DataGridViewElement.DataGridView> property was null. That behavior was unexpected and incorrect. These methods have been updated to simply return `false` if `DataGridView` is `null`.
```

**Review by lonitra — APPROVED** [review]
Overall looks good! 👍

**Review by BillWagner — APPROVED** [review]
This LGTM @gewarren 

Once you address the remaining comments, let's :shipit:

**Pull request merged**